source-repository versus source-repository-package

philderbeast opened the pull request:
Updates the users guide adding a warning comparing `source-repository` with `source-repository-package`, see #9655. Also;

- Adds pijul as `type`
- explicit `.` is the same as root

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] ~Is this a PR that fixes CI? If so, it will need to be backported to older cabal release branches (ask maintainers for directions).~

ulysses4ever commented on doc/cabal-package-description-file.rst:
```diff
@@ -2671,8 +2671,42 @@ Source Repositories
-repository for a package. Cabal lets you specify this information in
+.. warning::
+
+    A ``source-repository`` is for a version of the source of its package
```
I'm not a fan of the current version of this warning. My version:

Note that ``source-repository-package`` is not related to the ``source-repository`` field in package descriptions. [link the field]
The ``source-repository`` field specifies the source repository where upstream development of the package is happening. It is currently only used by ``cabal get`` via the same-named parameter if specified. [Link to the docs of the parameter] Whereas, the ``source-repository-package`` project stanza adds a local package to the project much like listing the package's source directory in the ``packages`` field, except that the source code is downloaded as needed instead of being taken from the local directory.

---

Also, the warning should go after the stanza has been introduced.

philderbeast replied:
> via the same-named parameter if specified

Sorry I don't quite follow @ulysses4ever. Do you mean `[PACKAGES]` by this?

```
$ cabal get --help
Download/Extract a package's source code (repository).

Usage: cabal get [PACKAGES]
...
```

philderbeast replied:
> Note that `source-repository-package` is not related to the `source-repository` field in package descriptions.

I know that these fields are both unrelated and related.

@ulysses4ever thanks for the suggested wording. I incorporated most of it.

ulysses4ever replied:
>> via the same-named parameter if specified
>
> Sorry I don't quite follow @ulysses4ever. Do you mean [PACKAGES] by this?


What I meant is the `-s` parameter that you can read about in `--help`:

```shellsession
❯ cabal get --help
Download/Extract a package's source code (repository).

Usage: cabal get [PACKAGES]

Creates a local copy of a package's source code. By default it gets the source
tarball and unpacks it in a local subdirectory. Alternatively, with -s it will
get the code from the source repository specified by the package.

Flags for get:
...
 -s[[head|this|...]], --source-repository[=[head|this|...]]
                                Copy the package's source repository (ie git
                                clone, darcs get, etc as appropriate).
```

I don't know about any other usage that that field in package descriptions has.

philderbeast replied:
Happy to add a mention of `cabal get` while I'm at it.

philderbeast replied:
I've done this, added a link to `cabal get`.

ulysses4ever reviewed (commented):
Thanks for the ping! Let's discuss!

philderbeast commented:
After taking a closer look at this, I went with adding a new page;

![image](https://github.com/haskell/cabal/assets/633283/a4db274c-67f3-4326-85ea-06ff6fc75adc)

philderbeast commented:
I corrected this disparity;

- type: token
- location: URL
- module: token
- branch: token
- tag: token
- subdir: directory

https://cabal.readthedocs.io/en/latest/cabal-package-description-file.html#source-repositories

- type: VCS kind
- location: VCS location (usually URL)
- tag: VCS tag
- subdir: subdirectory list
- post-checkout-command: command

https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#specifying-packages-from-remote-version-control-locations

ulysses4ever commented:
> After taking a closer look at this, I went with adding a new page. This way I could compare and contrast, be consistent and avoid duplication.

Having source repos at the same level as package description and project description seems wrong to me. It'd make more sense to me to add a page to CABAL GUIDE, e.g.: "How to utilize source repositories" or some such.

ulysses4ever commented on doc/cabal-package-description-file.rst:
```diff
@@ -2671,37 +2671,22 @@ Source Repositories
+   source control systems this involves specifying a tag, id or hash of some
+   form and perhaps a branch.
+
+As an example here are the repositories for the Cabal library. Note that the
```
```suggestion
As an example, here are the repositories for the Cabal library. Note that the
```

ulysses4ever commented:
Maybe I went ahead of myself here: the current version of that page still looks more like a reference + some commentary rather than a full-fledged guide. I think it'd be better to leave it as it was before, i.e. as an entry in the package/project description pages, and if you want to avoid duplication (a noble goal), just reference from one to another.

philderbeast commented:
> if you want to avoid duplication (a noble goal), just reference from one to another.

@ulysses4ever one is a set of `pkg-field` while the other is a set of `cfg-field` and I don't want to be messing with;

https://github.com/haskell/cabal/blob/3a275e5c7758372780fb1ad3247aa7c2d1297d31/doc/cabaldomain.py#L40

https://github.com/haskell/cabal/blob/3a275e5c7758372780fb1ad3247aa7c2d1297d31/doc/cabaldomain.py#L58

What we have is basically the same fields taken two ways and unless they're both on the same page, the duplication will go unnoticed and drift apart.

geekosaur commented on doc/cabal-package-description-file.rst:
```diff
@@ -2671,37 +2671,22 @@ Source Repositories
-Cabal supports specifying repositories for different use cases. By
-declaring which case we mean automated tools can be more useful. There
-are currently two kinds defined:
+There are two kinds and you can specify one or the other or both:
```
```suggestion
There are two kinds, and you can specify one or the other or both:
```

philderbeast replied:
Grammarly disliked the second clause so I made it clearer and then ended up splitting the sentence in two. Thanks for the suggestion @geekosaur.

Mikolaj commented:
Hi! What's the status here? @ulysses4ever: are you satisfied with the PR? To the point of approving or of asking me for the second review?

ulysses4ever commented:
@Mikolaj I am on the verge. My first concern is this change:

```
   cabal-package-description-file
   cabal-project-description-file
+  source-repositories
   cabal-config-and-commands
   external-commands
   setup-commands
```

I'm not sure it makes sense to have such a page on the top level.

Then, I don't know that this page fits well into the new top-level sectioning: Guide vs Reference. Phil puts it into reference but it has a good bit of discussion. Probably not enough to be called a guide, but also it's more prose than you'd expect from a reference page imo.

Finally, I don't know that it's a good idea to have one page for two things of which one belongs to .cabal files and one to project files. It feels wrong but maybe I just can't think out of the box.

In light of this, I'm abstaining from approval but I don't block.

I think the real reviewers for this PR are @malteneuss and @BinderDavid (in the light of #9214). Maybe if @andreabedini has a spare cycle...

philderbeast commented:
Might also help with #9816.

Mikolaj commented:
@malteneuss, @BinderDavid, @andreabedini, could you have a look at this PR?

malteneuss commented:
Hi @philderbeast, sorry for the late reply as i've been busy with my full-time job and a 2 year old, and thanks for taking the time to extending the docs.

Regarding the new structure i'm with @ulysses4ever on this one. 

> Having source repos at the same level as package description and project description seems wrong to me. It'd make more sense to me to add a page to CABAL GUIDE, e.g.: "How to utilize source repositories" or some such.

The package description reference ("cabal file") serves as a complete look-up list of all fields and values with their effects that you could add to a cabal file; here it would be wise to just add the missing values like `pijul`. If there would be duplication needed in the projection description file ("cabal.project") maybe a small reference/link to the "cabal file" section suffices; after all most users will likely use and look into "cabal files".

In general reference section should just describe "what is" or "what exists" with minimal explanation, "simple and to the point" (See https://documentation.divio.com/reference.html). 

Looking at your suggested titles like "As a Package Author" you may have indeed some specific cabal-user-oriented use cases and therefore a guide in mind along the lines: "How to add Haskell from other sources than Hackage?" 
So i would advise to move your main content to a new file `how-to-add-haskell-from-other-sources.rst` or similar (for SEO in the url) with a descriptive title as suggested above. I could then help to move on from there.

malteneuss commented on doc/source-repositories.rst:
```diff
@@ -0,0 +1,224 @@
+Source Repository Fields
+------------------------
+..
+  data SourceRepo = SourceRepo
+    { repoKind :: RepoKind
+    , repoType :: Maybe RepoType
+    , repoLocation :: Maybe String
+    , repoModule :: Maybe String
+    , repoBranch :: Maybe String
+    , repoTag :: Maybe String
+    , repoSubdir :: Maybe FilePath
+    }
```
I would advise against add this Haskell code, because as a user of cabal i'm not that interested in the underlying implementation. If you insist that this information is useful, maybe a link to the actual implementation from the refencre section would be good alternative.

philderbeast replied:
That code snippet is an RST comment and not rendered.

andreabedini commented:
Thanks for the ping but when it comes to documentation I defer to @malteneuss :relaxed: (I offer to entertain your 2yo in exchange).

malteneuss commented:
@philderbeast I see. Only having read your new `source-repositories.rst` file did i resolve some confusion (i think the one you want to resolve with your contribution, thanks) that `source-repository` and `source-repository-package` serve quite different purposes. The first is just metadata **provided** by a package author to locate the packages source code. The second is to actually **consume** Haskell packages from other than Hackage. 

So i propose you

1. turn  `source-repositories.rst` into a guide page with as consuming Haskell code from non-Hackage sources is surely a quite frequent use case that users should find quickly. Ideally, with a file name and title similar to e.g.  `how-to-add-haskell-from-other-sources.rst` or `how-to-consume-non-hackage-haskell-packages.rst`. I can review from there on.
2. restore the `source-repository` and `source-repository-package` sections but extended with your additional values that were missingm, and a clearer formulation of the purpose similar to your guide. Here a bit of duplication of the attributes is okay, as only `type`, `location` and `tag` are really identical to have the same short text. The other attributes need different formulations.
Both sections could then add a reference to your guide.

What do you think?

philderbeast commented:
> What do you think?

I'll try it, thanks for the suggestion @malteneuss.

philderbeast commented:
> as only `type`, `location` and `tag` are really identical to have the same short text

@malteneuss `branch` and `subdir` are also repeated, aren't they? For `source-repository-package` it is a list but for `source-repository`, `subdir` is one directory (a singleton list).

philderbeast commented:
@malteneuss could you please take another look?

I've added a new "How to source packages" guide page, renamed sections to use terms "package source" and "source packages" (or "source package dependencies"). I've also reworked the "As a Package Consumer" section. 

I've kept the "Version Control System Fields" help as a separate top-level page and put it just above the build info fields reference.

malteneuss commented:
Nice. I will take a look in the next days.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+Package Source
+--------------
+
+When we publish a package, its source is available from the package repository
+[#]_. The other way to access a package's source is from a source repository [#]_.
```
A higher level cabal-independent wording could help with context and clarity here. Probably something along the lines of
"When we publish a package, we can help users of our package to locate its source code by adding
package repository metadata [#]_ to our package. However, when we want use a package that is not, or not in the right version, available on Hackage, we can add that package that directly from its source repository (e.g. Git) [#]_."

philderbeast replied:
I'll have another go at this. I wanted to mention the two ways of getting source as a package consumer; 1) from hackage and 2) from a git repo.

![image](https://github.com/haskell/cabal/assets/633283/32d17f2b-465a-433e-98df-4928e37622d7)

philderbeast replied:
@malteneuss I've expand the introduction and added a section "Manually Obtaining Package *Source Code*" before talking about how cabal does this automatically.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+tools to interpret and make effective use of this information.
+
+- ``source-repository`` says where to find the source for a **package description**
+- ``source-repository-package`` locates source for **project dependencies**
```
Better "locates the source code to be added as a **project dependency**"

philderbeast replied:
I went with this suggestion, change the list to be a description list and added a warning about the confusing names.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+code of each dependency and will download it as needed. Example of commands
+like this are ``cabal build`` or ``cabal freeze``.
+
+Dependencies of packages in a project are sourced, by default, from the hackage
```
better use a more common wording like "project are downloaded". Sourced feels only familiar to experienced terminal users.

philderbeast replied:
I opted to use "sourced" in the everyday sense as in obtain or get.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+like this are ``cabal build`` or ``cabal freeze``.
+
+Dependencies of packages in a project are sourced, by default, from the hackage
+package repository. Cabal will download the source for these from hackage.
```
Better "the source code and ..." something else you want to mention if you indeed want to emphasize something here, otherwise it's redundant to the sentence before.

philderbeast replied:
I started putting *source code* in italics. I hope that helps. I've changed this to;

```rst
Dependencies of packages in a project are sourced, by default, from the hackage
package repository. Cabal will download the *source code* ``.tar.gz`` archive
for these from hackage.
```

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+:ref:`package source<pkg-author-source>` you will be dealing with
+``source-repository``, a package description stanza that typically specifies
+where upstream development of the package is happening.  A package published to
+hackage will be displayed with a link to the source repository if a
```
Hackage (uppercase)

philderbeast replied:
Thanks, all uppercase now.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+package repository. Cabal will download the source for these from hackage.
+
+We can also take dependencies on :ref:`source packages<pkg-consume-source>`
+that are accessed via a version control system (VCS) such as git and set up in
```
Git (uppercase)

philderbeast replied:
Thanks for noticing. Corrected.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+
+We can also take dependencies on :ref:`source packages<pkg-consume-source>`
+that are accessed via a version control system (VCS) such as git and set up in
+the project with ``source-package-repository`` stanzas. With these, we can take
```
stanzas ->  sections. stanzas (although technically correct) is less familiar and understandable than section. Even better to add a reference to the cabal project description-file section.

philderbeast replied:
Seems like we're using "section" in the guide and "stanza" in the reference, aren't we?

philderbeast replied:
I added a link but kept "stanza".

```
We can also :ref:`take a dependency from a source code
repository<pkg-consume-source>`. These are accessed via a version control system
(VCS) such as Git and set up in the project with ``source-package-repository``
stanzas.
```

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+the project with ``source-package-repository`` stanzas. With these, we can take
+dependencies on packages not published to hackage, on revisions not yet
+published or on forks.  This is the easiest way to work with a fork, much
+easier than using a git submodule.  A dependency taken this way, effectively
```
Git (uppercase)

philderbeast replied:
Thanks again, fixed.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+directory in the packages field, except that the source code is downloaded by
+cabal using the version control system.
+
+.. [#] Each package on hackage is available for download as a ``.tar.gz``
```
Hackage (uppercase)

philderbeast replied:
Footnote moved to note and casing fixed.

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+The location of the repository, usually a URL but the exact form of this field
+depends on the repository type. For example:
+
+-  for darcs: ``http://code.haskell.org/foo/``
+-  for git: ``git://github.com/foo/bar.git``
+-  for CVS: ``anoncvs@cvs.foo.org:/cvs``
```
Upper case for project names.

philderbeast replied:
Done.

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+
+Many source control systems support the notion of a branch, as a distinct
+concept from having repositories in separate locations. For example CVS, SVN and
+git use branches while darcs uses different locations for different branches. If
```
Git, Darcs ...

philderbeast replied:
Done.

malteneuss commented on doc/cabal-project-description-file.rst:
```diff
@@ -205,8 +205,10 @@ Formally, the format is described by the following BNF:
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. _pkg-consume-source:
+
+Source Package Dependencies
```
The older title was more understandable to people unfamiliar with what a source package is, but probably a bit too long. Maybe "Dependencies from Remote Version Control Systems" ?

philderbeast replied:
Changed the title to "Taking a Dependency from a Source Code Repository", a bit long but not too long.

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+-  ``mercurial`` (or alias ``hg``)
+-  ``bazaar`` (or alias ``bzr``)
+-  ``arch``
+-  ``monotone``
+-  ``pijul``
+
+The VCS kind will determine what other fields are appropriate to specify for a
+particular version control system.
+
+VCS location
+^^^^^^^^^^^^
+
+The location of the repository, usually a URL but the exact form of this field
+depends on the repository type. For example:
+
+-  for darcs: ``http://code.haskell.org/foo/``
+-  for git: ``git://github.com/foo/bar.git``
+-  for CVS: ``anoncvs@cvs.foo.org:/cvs``
+
+VCS branch
+^^^^^^^^^^
+
+Many source control systems support the notion of a branch, as a distinct
+concept from having repositories in separate locations. For example CVS, SVN and
+git use branches while darcs uses different locations for different branches. If
+you need to specify a branch to identify a your repository then specify it in
+this field.
+
+VCS tag
+^^^^^^^
+
+A tag identifies a particular state of a source repository.  The exact form of
+the tag depends on the repository type.
+
+VCS subdirectory
+^^^^^^^^^^^^^^^^
+
+A field of this type is always optional because it defaults to empty, which
+corresponds to the root directory of the repository and is the same as
+specifying ``.`` explicitly.
```
These explanations paragraphs better belong to the field descriptions below. 

Also i'm still not convinced it makes sense to create a dedicated reference page to factor out the commonalities as the two use cases are rather independent (when i add a source repository stanza to add upstream metadata, i don't want to immediately add a VCS dependency, and vice versa). I would rather go for a bit of text duplication in the field descriptions to be able to read each stanza description on its own.
@ulysses4ever @andreabedini What do you think?

philderbeast replied:
@malteneuss how do you feel about this now, now that redirects to it describe their fields?

malteneuss replied:
It's better, yes, and good enough that i won't block this.

malteneuss commented on doc/cabal-package-description-file.rst:
```diff
@@ -2676,43 +2676,31 @@ Starting with Cabal-2.2 it's possible to use common build info stanzas.
-Cabal supports specifying different information for various common
-source control systems. Obviously not all automated tools will support
-all source control systems.
+Where to find the source for this package.
```
better "source code"

philderbeast replied:
Expanded it to "Marks out where to find the *source code* for this package within a **source code repository**."

malteneuss commented on doc/cabal-package-description-file.rst:
```diff
@@ -2676,43 +2676,31 @@ Starting with Cabal-2.2 it's possible to use common build info stanzas.
+
+.. _pkg-author-source:
+
+Package Source
```
better "Package Source Metadata", "Package Source Code Origin", "Locate Package Source Code" or similar to better reveal the intent.

philderbeast replied:
Went with "Source Marker".

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+corresponds to the root directory of the repository and is the same as
+specifying ``.`` explicitly.
+
+Some projects put the sources for multiple packages under a single source
```
Better "..multiple packages inside a single VCS repository".

philderbeast replied:
Thanks, I used your suggestion exactly.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+   source archive. For example, here's the direct link to download
+   `vector-0.13.1.0.tar.gz <vector-download>`_.
+
+.. [#] Also known as a source control repository or version control system
+   (VCS) repository.
+
+.. _vector-download: https://hackage.haskell.org/package/vector-0.13.1.0/vector-0.13.1.0.tar.gz
```
Maybe you could add an example how to add such a tar.gz file as a package dependency?
And also an example using Git would be good as this is the most common VCS

philderbeast replied:
Yes, thanks for suggesting adding these examples.

philderbeast replied:
I've added a whole vendoring section about this.

malteneuss reviewed (changes requested):
I really like the initiative to have a dedicated guide page on this topic with your easy to understand formulations. However, i'm less inclined to have a separate reference page that mixes two different use cases. (I haven't checked formatting yet; will do that when structure is settled)

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,65 @@
+How to source packages
```
I'd like to keep his guide about everything source. I hope the expanded introduction helps.

philderbeast replied:
I change the guide title to "How to get package *source code*".

philderbeast commented:
@malteneuss thanks for reviewing the `doc/version-control-fields.rst` file. A lot of the content there I didn't write but am happy to improve with your help

malteneuss commented on doc/index.rst:
```diff
@@ -13,6 +13,7 @@ Welcome to the Cabal User Guide
    :numbered:
    :maxdepth: 2
 
+   how-to-source-packages
```
Please move the new guid to the second position as the usual dependency handling is much more frequent.

philderbeast replied:
:+1:

malteneuss replied:
Maybe the commit got lost; it's still showing 
`how-to-source-packages` before `how-to-package-haskell-code`

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+    :header-rows: 1
+    :widths: 30 30 40
+
+    * - Field Name
+      - source-repository (head|this)
+      - source-repository-package
+    * - type
+      - 🗹
+      - 🗹
+    * - location
+      - 🗹
+      - 🗹
+    * - branch
+      - 🗹
+      - 🗹
+    * - tag
+      - 🗹
+      - 🗹
+    * - subdir
+      - 🗹 (0 or 1)
+      - 🗹 (0 or 1 for each dependency)
+    * - module (CVS only)
+      - 🗹
+      - ☐
+    * - post-checkout-command
+      - ☐
+      - 🗹
```
The symbols look broken on my Linux machine in Firefox.
![version_control_fields](https://github.com/haskell/cabal/assets/5301202/30016a50-8ae5-4369-a417-6cd536a6a067)

philderbeast replied:
Renders fine for me on Ubuntu with Firefox. What locale are you using @malteneuss? Is the problem there if you use the link from the CI rendering, from the "docs/readthedocs.org:cabal" job (might need to scroll down to see the link).

geekosaur replied:
They're also broken for me in the RTD documentation. Ubuntu 22.04.3, Chrome.

philderbeast replied:
Do code blocks like the following with unicode chars render well? These snippets are from `doc/how-to-source-packages.rst`.

```
       (package description)      source-repository         (Git repository)
      solver-benchmarks.cabal ──────────────────────────⮚ github/haskell/cabal

        (project)       source-repository-package         (Git repository)
      cabal.project ──────────────────────────────────⮚ github/haskell/vector
```

geekosaur replied:
It appears to have half-rendered: the lines are fine but there's a small block at the end which may have been intended to be an arrowhead?

geekosaur replied:
![Screenshot_2024-05-02_13-29-20](https://github.com/haskell/cabal/assets/2036384/d220f326-f5e1-45bd-88c0-5db05d74b0b6)

philderbeast replied:
Yes, the lines I grabbed from the output from the `tree` command in ubuntu and the arrow from https://www.amp-what.com/unicode/search/arrowhead.

philderbeast replied:
Thanks for the screenshot @geekosaur. Unfortunate that those unicode chars don't work. In the past I've had good results using them in web pages.

malteneuss replied:
Then let's just use `---->` ascii arrows and ascii symbols if possible for now. We can try out beatifications later

philderbeast replied:
As it is only the arrowheads, I replaced those and the ticked checkboxes.

malteneuss commented on doc/version-control-fields.rst:
```diff
@@ -0,0 +1,179 @@
+Version Control System Fields
+=============================
+
+Most of the version control system (VCS) fields types are common to both
+``source-repository`` and ``source-repository-package`` stanzas.
```
This `version-control-fields.rst` page feels out of place compared to the other reference pages:
The others summarize a high-level thing like the whole file formats or a major part of the CLI interface. This one picks and focuses on two minor stanzas, and lacks context (page can't be read on its own).
If we decide to keep this page (which i strongly discourage, but won't block if others are fine with it), it should have a context introductory sentence that links to the two other reference pages.

philderbeast replied:
This page now only has VCS fields. The `source-repository` is back with `doc/cabal-package-description-file.rst` and the `source-repository-package` is back with `doc/cabal-project-description-file.rst`.

malteneuss commented on doc/index.rst:
```diff
@@ -29,6 +30,7 @@ Welcome to the Cabal User Guide
    external-commands
    setup-commands
    file-format-changelog
+   version-control-fields
```
This page's position better fits right after the two cabal description files as it strongly belong to them

philderbeast replied:
Thanks, I'd put it next to the "Field Syntax Reference" but have moved it as you suggest.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,108 @@
+package repository. Cabal will download the *source code* ``.tar.gz`` archive
+for these from hackage.
+
+We can also :ref:`take a dependency from a source code
+repository<pkg-consume-source>`. These are accessed via a version control system
```
works

malteneuss reviewed (commented):
Minor final remarks after taking a look the whole documentation website.
A final decision is required if we continue with the separate reference page for version control fields.

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,108 @@
+repository<pkg-consume-source>`. These are accessed via a version control system
+(VCS) such as Git and set up in the project with ``source-package-repository``
+stanzas. With these, we can take dependencies on packages not published to
+hackage, on revisions not yet published or on forks.  This is the easiest way to
```
:+1:

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,108 @@
+work with a fork, much easier than using a Git submodule.  A dependency taken
+this way, effectively adds a local package to the project much like listing the
+package's source directory in the packages field, except that the source code is
+downloaded by cabal using the version control system.
```
:+1:

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,108 @@
+code of each dependency and will download it as needed. Example of commands
+like this are ``cabal build`` or ``cabal freeze``.
+
+Dependencies of packages in a project are sourced, by default, from the hackage
```
:+1:

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+
+You can use ``cabal get`` and then add the unpack directory to the ``packages``
+field in the project's ``cabal.project`` file.  This downloads the package's
+``.tar.gz`` source code archive and extracts it to directory named after the
```
```suggestion
``.tar.gz`` source code archive and extracts it to a directory named after the
```

philderbeast replied:
:+1:

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+Vendoring unpacked *source code*
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can use ``cabal get`` and then add the unpack directory to the ``packages``
```
A ref to cabal get section would be good

philderbeast replied:
:+1:

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+      $ echo "packages: vector-0.13.1.0" >> cabal.project
+
+You can vendor unpacked source code obtained by other means, by means that
+replace the ``cabal unpack`` step, in the same way.
```
There's no `cabal unpack` call here.

philderbeast replied:
Thanks. I went with;

```rst
   .. code-block:: shell

      $ cabal update
      $ cabal get vector
      Unpacking to vector-0.13.1.0/
      $ echo "packages: vector-0.13.1.0" >> cabal.project

You can vendor unpacked source code obtained by other means, by means that
replace the ``cabal get`` unpacking step, in the same way.
```

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+You can vendor unpacked source code obtained by other means, by means that
+replace the ``cabal unpack`` step, in the same way.
+
+Fork don't vendor
```
```suggestion
Fork, don't vendor
```

philderbeast replied:
:+1:

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+         location: https://github.com/haskell/vector.git
+         subdir:   vector
+
+   To turn that into a dependency we'd add the following to our project, adding
+   a ``tag`` field if we wanted to pin to a particular commit;
+
+   .. code-block:: cabal
+
+      source-repository-package
+         type:     git
+         location: https://github.com/haskell/vector.git
+         subdir:   vector
+
```
It would be good to add a tag field to the example to point to the 0.13.1.0 version and continue the theme of using that version throughout the  guide page

philderbeast replied:
Can do now that I see there's a tag for it. This version is not in [github/vector releases](https://github.com/haskell/vector/releases) but there is a [vector-0.13.1.0 tag](https://github.com/haskell/vector/tree/vector-0.13.1.0).

philderbeast replied:
I've include the tag field in the example and shown how I got that commit hash, with;

```
$ git ls-remote --tags
From git@github.com:haskell/vector.git
...
79bdd2edcfaf6b07f7fabc43a7d9c5a2ff93d3ca     refs/tags/vector-0.13.1.0
```

I also moved this content, that was in a note, up to a subsection. I gave it the title of "Setting up a source code dependency".

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+How to get package *source code*
```
```suggestion
How to use package *source code*
```

philderbeast replied:
I went with "How to locate and get package *source code* and reworked the introductory two sentences of that section.

philderbeast replied:
Thought that "locate and get" was too long. Considered "muster" but the meaning I'm familiar with is NZ/AU only so went with "shepherd".

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+
+We can **"source"** or obtain the package *source code* and we can leave a
+**"source"** marker in the package description that shows where to find the
+*source code*. This marker is the ``source-repository`` field and is described
+in the :ref:`package author<pkg-authors>` section.
```
As this paragraph gives an overview over the guide, a sentence about `source-repository-package` and a ref to _pkg-consumers would be good (as it was before)

philderbeast replied:
I've added back the ref for this.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+
+.. _pkg-consumers:
+
+*Source code* when taking a package dependency
```
```suggestion
Taking *Source code* as a package dependency
```

philderbeast replied:
I disagree with this change of section title. The section talks about how Cabal deals with source code as well. The section starts with this before it goes on to talk about how to deal with dependency source code ourselves.

philderbeast replied:
@malteneuss if you agree, can you please resolve this conversation?

malteneuss replied:
I can't due to https://github.com/orgs/community/discussions/8327 xD

philderbeast replied:
OK, I can mark it as resolved.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+         build-depends:
+      ...
+            vector
+      ...
```
```suggestion
           ...
```

philderbeast replied:
I removed that line by running dependencies together in one line, `build-depends: base, ..., vector`.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+      ...
+      library
+         build-depends:
+      ...
```
better remove this `...`

philderbeast replied:
I'm eliding dependencies, not wanting to show them all. I was able to get it done in fewer lines by changing to;

```
library
   build-depends: base, ..., vector
```

Not verbatim from the package description but close enough.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+
+.. Warning::
+
+   The ``source-repository`` field, the one without "package" is for a package!
```
```suggestion
   The ``source-repository`` field, the one without a "-package" suffix, belongs to a package!
```

philderbeast replied:
:+1:

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+            vector
+      ...
+
+   The ``source-repository-package`` field, the one with "package" isn't for a
```
```suggestion
   The ``source-repository-package`` field, the one with a "-package" suffix, doesn't belong to a
```

Is that correct though? I always thought you could add this field to a `*.cabal` file as well.

geekosaur replied:
No, it's `cabal.project` only.

philderbeast replied:
I've made the change to match the suggestion.

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+
+   The ``source-repository-package`` field, the one with "package" isn't for a
+   package! It's for a project dependency. This name is a bit misleading because
+   it can describe multiple dependencies and not just one. That it is for
```
```suggestion
   it can describe multiple package dependencies and not just one. That it is for
```

philderbeast replied:
I went with;

```rst
   The ``source-repository-package`` field, the one with a "-package" suffix,
   doesn't belong to a package! It's for a project dependency. This name is a
   bit misleading being singular because it can describe multiple dependency
   packages at the same location, each offset by a directory containing a
   package description, a ``.cabal`` file.
 ```

malteneuss commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,309 @@
+   The ``source-repository-package`` field, the one with "package" isn't for a
+   package! It's for a project dependency. This name is a bit misleading because
+   it can describe multiple dependencies and not just one. That it is for
+   "package" also makes sense. Ultimately it is the union of all dependencies of
+   local packages within the project that actually declare the set of project
+   dependencies, not the project.  The project can tighten version constraint
+   ranges with ``constraints`` or loosen them with ``allow-newer`` or
+   ``allow-older`` but it cannot add or remove dependencies.
```
This paragraph is hard to follow without an example and a clearer distinction between my local project and the remote/external dependency repository.

philderbeast replied:
I made this a separate warning and simplified it.

malteneuss reviewed (commented):
Really nice examples. Now you get a much better understanding what's possible with Cabal and source code.

philderbeast commented:
@malteneuss I've reread the rendered help and it is ready for another review.

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,342 @@
+How to shepherd package *source code*
```
I'm using https://www.thesaurus.com/browse/guide for inspiration. Maybe "handle" covers all the cases; marking, getting, redirecting etc, well enough.

philderbeast replied:
Went with "deal with".

malteneuss commented:
LGTM

malteneuss commented on doc/index.rst:
```diff
@@ -13,6 +13,7 @@ Welcome to the Cabal User Guide
    :numbered:
    :maxdepth: 2
 
+   how-to-source-packages
    how-to-package-haskell-code
```
```suggestion
   how-to-package-haskell-code
   how-to-source-packages
```

malteneuss replied:
Last nitpick, then LGTM

philderbeast replied:
Done, thanks.

philderbeast commented:
Once approved, I'll squash the commits.

malteneuss commented:
Apparently, a PR needs two approvals by people with merge rights

philderbeast commented:
@geekosaur thanks for approving this once before. It has since had a lot of rework.

geekosaur commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,342 @@
+Fork, don't vendor
+^^^^^^^^^^^^^^^^^^
+There's no need to vendor packages on Hackage if you expect Hackage to always be
+available as packages cannot be deleted from Hackage. Source code repositories,
```
```suggestion
available, as packages cannot be deleted from Hackage. Source code repositories,
```

philderbeast replied:
:+1:

geekosaur reviewed (approved):
Minor nits

philderbeast commented on doc/how-to-source-packages.rst:
```diff
@@ -0,0 +1,342 @@
+
+.. Warning::
+
+   Cabal project, ``cabal.project``, files don't declare dependencies!
```
:+1:

philderbeast commented:
Label merge+no rebase is necessary when the pull request is from an organisation.

philderbeast commented:
@Mikolaj I don't understand what state this PR is in. The page is showing "1 approval" but two reviewers have approved it. Don't ask me why one tick is green but the other grey, next to the reviewers. Hovering over each tick, the tip says the reviewer has approved it. Time to review the review approvals, I guess?

Mikolaj commented:
Well spotted. I've fixed this by bumping @malteneuss's authority from Triage to Write.

Mikolaj commented:
@mergify refresh

Mikolaj commented:
@mergify rebase

Mikolaj commented:
Let me add the `delay_passed` label manually, because it should be auto-assigned by mergify already.

ulysses4ever commented:
I still don't get why we have a "version control fields" page at the top level. (Available now as https://cabal.readthedocs.io/en/latest/ CABAL REFERENCE / 3)

philderbeast commented:
@ulysses4ever short summaries (required etc) in the following:

* [Source code repository marker](https://cabal.readthedocs.io/en/latest/cabal-package-description-file.html#source-code-repository-marker) (package use of these fields)

* [Taking a dependency from a source code repository](https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#taking-a-dependency-from-a-source-code-repository) (project use of these fields)

Details of these types of fields in:

* [Version Control System Fields](https://cabal.readthedocs.io/en/latest/version-control-fields.html)

malteneuss commented:
I didn't want to stand in the way as it gives more exposure to the fact that Cabal can use repositories raw without Hackage.
However, i'm also not sure if a top-level page is the best place to put this information vs. a bit of duplication. Of course the current setup is not set in stone...

geekosaur commented:
It really wants to get its own section in the Field Syntax Reference, I think.

ulysses4ever commented:
@BinderDavid you redesigned the top-level sectioning of the Cabal Manual. Can you look into this PR and suggest a better place for the new page that currently goes under CABAL REFERENCE and talks about version control systems (or confirm that the current setup is the best possible in your opinion)?

BinderDavid commented:
Sure. I am at Zurihac right now, and I can take a look tomorrow. I also have some unfinished cabal user guide PRs that I wanted to finish anyway, so that's a good occasion to pick them up again ^^
